File: ExamplesConsole.VB/ExampleHelper.vb
' Copyright 2009-2014 Marc Fletcher, Matthew Dean
' 
' This program is free software: you can redistribute it and/or modify
' it under the terms of the GNU General Public License as published by
' the Free Software Foundation, either version 3 of the License, or
' (at your option) any later version.
' 
' This program is distributed in the hope that it will be useful,
' but WITHOUT ANY WARRANTY; without even the implied warranty of
' MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
' GNU General Public License for more details.
' 
' You should have received a copy of the GNU General Public License
' along with this program.  If not, see <http://www.gnu.org/licenses/>.
' 
' A commercial license of this software can also be purchased. 
' Please see <http://www.networkcomms.net/licensing/> for details.

Imports NetworkCommsDotNet
Imports NetworkCommsDotNet.Tools

Module ExampleHelper
    Private lastServerIPEndPoint As IPEndPoint = Nothing

    Function GetServerDetails() As ConnectionInfo
        If Not IsNothing(lastServerIPEndPoint) Then
            Console.WriteLine("Please enter the destination IP and port. To reuse '{0}:{1}' use r:", lastServerIPEndPoint.Address, lastServerIPEndPoint.Port)
        Else
            Console.WriteLine("Please enter the destination IP address and port, e.g. '192.168.0.1:10000':")
        End If

        While True
            Try
                'Parse the provided information
                Dim userEnteredStr As String = Console.ReadLine()

                If userEnteredStr.Trim() = "r" And Not IsNothing(lastServerIPEndPoint) Then
                    Exit While
                Else
                    lastServerIPEndPoint = IPTools.ParseEndPointFromString(userEnteredStr)
                    Exit While
                End If
            Catch ex As Exception
                Console.WriteLine("Unable to determine host IP address and port. Check format and try again:")
            End Try
        End While

        Return New ConnectionInfo(lastServerIPEndPoint)
    End Function
End Module

File: Platforms/Net80/ExamplesConsole/ExampleHelper.cs
// 
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
// 
//   http://www.apache.org/licenses/LICENSE-2.0
// 
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.
// 

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using NetworkCommsDotNet;
using System.Net;
using NetworkCommsDotNet.Connections;
using NetworkCommsDotNet.Tools;

namespace Examples.ExamplesConsole
{
    /// <summary>
    /// Provides implementation shared across examples
    /// </summary>
    public static class ExampleHelper
    {
        static IPEndPoint lastServerIPEndPoint = null;

        /// <summary>
        /// Request user to provide server details and returns the result as a <see cref="ConnectionInfo"/> object. Performs the necessary validation and prevents code duplication across examples.
        /// </summary>
        /// <param name="applicationLayerProtocol">If enabled NetworkComms.Net uses a custom 
        /// application layer protocol to provide useful features such as inline serialisation, 
        /// transparent packet transmission, remote peer handshake and information etc. We strongly 
        /// recommend you enable the NetworkComms.Net application layer protocol.</param>
        public static ConnectionInfo GetServerDetails(ApplicationLayerProtocolStatus applicationLayerProtocol = ApplicationLayerProtocolStatus.Enabled)
        {
            if (lastServerIPEndPoint != null)
                Console.WriteLine("Please enter the destination IP and port. To reuse '{0}:{1}' use r:", lastServerIPEndPoint.Address, lastServerIPEndPoint.Port);
            else
                Console.WriteLine("Please enter the destination IP address and port, e.g. '192.168.0.1:10000':");

            while (true)
            {
                try
                {
                    //Parse the provided information
                    string userEnteredStr = Console.ReadLine();

                    if (userEnteredStr.Trim() == "r" && lastServerIPEndPoint != null)
                        break;
                    else
                    {
                        lastServerIPEndPoint = IPTools.ParseEndPointFromString(userEnteredStr);
                        break;
                    }
                }
                catch (Exception)
                {
                    Console.WriteLine("Unable to determine host IP address and port. Check format and try again:");
                }
            }

            return new ConnectionInfo(lastServerIPEndPoint, applicationLayerProtocol);
        }
    }
}
